worker/src/handlers/aggregationProcessor/aggregationProcessor.js

```diff
@@ -0,0 +1,222 @@
+import { LOCK_TIMEOUT_MINUTES, AGGREGATION_PROCESSOR_QUEUE } from 'lib/constants/aggregationProcessor';
+import moment from 'moment';
+import AggregationProcessor from 'lib/models/aggregationProcessor';
+import Statement from 'lib/models/statement';
+import { get } from 'lodash';
+import { publish } from 'lib/services/queue';
+
+export const combine = (as, bs, keyFn, mergeFn) => {
+  const groupedAs = as.reduce((result, a) => {
+    const key = keyFn(a);
+    result[key] = { a };
+    return result;
+  }, {});
+  const groupedResults = bs.reduce((result, b) => {
+    const key = keyFn(b);
+    result[key] = { ...result[key], b };
+    return result;
+  }, groupedAs);
+  return Object.entries(groupedResults).map(([model, results]) =>
+    mergeFn(model, results.a, results.b)
+  );
+};
+
+export const hasReachedEnd = ({ model, now }) => {
+  if (
+    moment(model.fromTimestamp).isSameOrBefore(moment(model.gtDate)) &&
+    moment(model.toTimestamp).isSame(now)
+  ) {
+    return true;
+  }
+  return false;
+};
+
+export const getFromTimestamp = ({ model, now }) => {
+  const blockSizeSeconds = moment(now).subtract(model.blockSizeSeconds, 'seconds');
+  const windowSize = moment(now).subtract(model.windowSize, model.windowSizeUnits);
+
+  if (blockSizeSeconds.isAfter(windowSize)) {
+    return blockSizeSeconds;
+  }
+
+  return windowSize;
+};
+
+export const getAddFromTimestamp = ({ model, now }) => {
+  if (model.toTimestamp) {
+    return moment(model.toTimestamp);
+  }
+
+  return getFromTimestamp({ model, now });
+};
+
+/*
+  returns undefined from to now is less than the window size
+*/
+export const getAddToTimestamp = ({
+  model,
+  now
+}) => {
+  const addFrom = getAddFromTimestamp({ model, now });
+  const nextAddFrom = addFrom.add(model.blockSizeSeconds, 'seconds');
+
+  if (nextAddFrom.isSameOrAfter(now)) {
+    return now;
+  }
+  return nextAddFrom;
+};
+
+const getAddPipeline = ({
+  model,
+  now
+}) => {
+  const addPipeline = [
+    { $match: {
+      timestamp: {
+        $gt: getAddFromTimestamp({ model, now }).toDate(),
+        $lte: getAddToTimestamp({ model, now }).toDate()
+      }
+    } },
+    ...(JSON.parse(model.pipelineString))
+  ];
+
+  return addPipeline;
+};
+
+const getSubtractPipeline = ({
+  model,
+  now
+}) => {
+  const hasSubtraction = model.fromTimestamp || model.toTimestamp;
+  if (!hasSubtraction) {
+    return;
+  }
+
+  const fromTimestamp = getFromTimestamp({ model, now });
+
+  const subtractPipeline = [
+    {
+      $match: {
+        timestamp: {
+          $gte: fromTimestamp.toDate(),
+          $lt: moment(model.fromTimestamp).toDate()
+        }
+      }
+    },
+    ...(JSON.parse(model.pipelineString))
+  ];
+
+  return subtractPipeline;
+};
+
+const aggregationProcessor = async ({
+  aggregationProcessorId,
+  publishQueue = publish,
+}, done) => {
```

Comment:
I don't think we need a `done` callback since this is async. I think we should avoid having callbacks where possible.